pallets/xyk/src/lib.rs

```diff
@@ -853,28 +795,13 @@ impl<T: Config> AMM<T::AccountId, AssetId, AssetPair, Balance> for Pallet<T> {
 		max_limit: Balance,
 		discount: bool,
 	) -> Result<AMMTransfer<T::AccountId, AssetId, AssetPair, Balance>, DispatchError> {
-		ensure!(
-			amount >= T::MinTradingLimit::get(),
-			Error::<T>::InsufficientTradingAmount
-		);
-
-		ensure!(Self::exists(assets), Error::<T>::TokenPoolNotFound);
+		Self::do_validate_buy_common(assets, amount)?;
 
 		let pair_account = Self::get_pair_id(assets);
 
 		let asset_out_reserve = T::Currency::free_balance(assets.asset_out, &pair_account);
```

Comment:
I wonder how expensive to call `Self::get_pair_id()` and `T::Currency::free_balance()`. Because they are already called in the validate method. If they are expensive to call, I could return them from the `do_validate...()` method.

Reply:
it is cheap, and this is probably OT for this PR anyway

Reply:
Should be considered as memory read if it's read once b4. I would be interested if the benchmarking correctly assumes this

Reply:
all right, I will leave it like that if it is a memory read, so it should be cheap and fast.

As for benchmarking, I assume that it takes this into account, as the benchmark tests are executed against the runtime with real dependencies. Otherwise, it would be strange, in my opinion.